py/desiutil/annotate.py

```diff
@@ -0,0 +1,376 @@
+        for s in search:
+            if s in key.lower():
+                return key
+    raise KeyError(f"No key matching '{search[0]}' found!")
+
+
+def yml_units(filename):
+    """Parse a YAML file that contains column names and units and optionally comments.
+
+    The YAML file should contain a dictionary with a keyword like 'units' and,
+    optionally, a keyword like 'comments'.
+
+    For backwards-compatibility, the YAML file can be simply a dictionary
+    containing column names.
+
+    Parameters
+    ----------
+    filename : :class:`str` or :class:`pathlib.Path`
+        Read column definitions from `filename`.
+
+    Returns
+    -------
+    :class:`tuple`
+        A tuple containing two :class:`dict` objects for units and comments.
+        If no comments are detected, the comments :class:`dict` will be empty.
+    """
+    comments = dict()
+    # log.debug("y = yaml.safe_load('%s')", filename)
+    with open(filename, newline='') as f:
+        y = yaml.safe_load(f)
+    try:
+        u = yml_unit_key(y)
+    except KeyError:
+        log.warning(f"{filename} does not have a unit column, assuming keys are columns!")
+        u = None
+    try:
+        c = yml_unit_key(y, comment=True)
+    except KeyError:
+        c = None
+    if u:
+        units = y[u]
+    else:
+        units = y
+    if c:
+        comments = y[c]
+    return (units, comments)
+
+
+def annotate_table(table, units, comments=None, inplace=False):
+    """Add annotations to `table`.
+
+    Parameters
+    ----------
+    table : :class:`astropy.table.Table` or :class:`astropy.table.QTable`
+        A data table.
+    units : :class:`dict`
+        Mapping of table columns to units.
+    comments : :class:`dict`, optional
+        Mapping of table columns to comments.
+    inplace : :class:`bool`, optional
+        If ``True``, modify `table` directly instead of returning a copy.
+
+    Returns
+    -------
+    :class:`astropy.table.Table`
+        An updated version of `table`.
+    """
+    if inplace:
+        t = table
+    else:
+        if isinstance(table, QTable):
+            t = QTable(table)  # copy=True is the default.
+        else:
+            t = Table(table)
+    for column in units:
+        if column in t.colnames:
+            if len(units[column]) > 0:
+                try:
+                    log.debug("t['%s'].unit = '%s'", column, units[column])
+                    t[column].unit = units[column]
+                except AttributeError:
+                    #
+                    # Can't change .unit if it is already set. Try to convert.
+                    #
+                    try:
+                        log.debug("t.replace_column('%s', t['%s'].to('%s'))", column, column, units[column])
+                        t.replace_column(column, t[column].to(units[column]))
+                    except UnitConversionError:
+                        log.error("Cannot add or replace unit '%s' to column '%s'!", units[column], column)
```

Comment:
consider adding an equivalent units check to `csv_units` and `yml_units` to catch invalid units at the time they are loaded, before attempting to apply them to a table.

OTOH, it if the input file has invalid units for a column that isn't in the table anyway, it could be considered a feature to not complain until you actually try to use it, so I could see an argument either way.

Reply:
Right, but that would come into play when implementing the command-line script that adds units to FITS files, as opposed to a `Table`.

However, I should emphasize here that the `UnitConversionError` exception isn't checking for the validity of the unit *format*. That is catching the much more specific case where *existing* units are being *converted* to other units requested by the user, for example, `nm` --> `Angstrom` or `deg` --> `arcsec`.

Still, I agree that there should be a check on the validity of the unit *format*.